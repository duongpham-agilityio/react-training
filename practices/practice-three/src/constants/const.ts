export const LIMIT_QUANTITY: number = 5;
