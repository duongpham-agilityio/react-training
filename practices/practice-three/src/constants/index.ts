export * from './url';
export * from './message';
export * from './variable';
export * from './title';
export * from './regex';
export * from './time';
export * from './status';
