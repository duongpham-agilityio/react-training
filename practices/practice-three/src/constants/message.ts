interface IMessages {
  EMPTY: string;
  FAIL_TO_FETCH: string;
  ADD_TO_CART_SUCCESS: string;
  ADD_TO_CART_FAIL: string;
  REMOVE_FORM_CART_SUCCESS: string;
  REMOVE_FORM_CART_FAIL: string;
}

export const MESSAGES: Readonly<IMessages> = {
  EMPTY: 'No data',
  FAIL_TO_FETCH: 'Something went wrong!! Please reload page.',
  ADD_TO_CART_SUCCESS: 'Add to cart successful!',
  ADD_TO_CART_FAIL: 'Add to cart fail!',
  REMOVE_FORM_CART_SUCCESS: 'Remove from cart successful!',
  REMOVE_FORM_CART_FAIL: 'Remove from cart fail!',
};
