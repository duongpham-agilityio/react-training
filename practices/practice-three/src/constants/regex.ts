export const REGEXPS = {
  EMAIL: /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/,
};
