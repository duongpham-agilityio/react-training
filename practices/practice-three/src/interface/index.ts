export * from './cart';
export * from './account';
export * from './product';
