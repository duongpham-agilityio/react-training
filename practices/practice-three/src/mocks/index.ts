export * from './productCard';
export * from './filterBar';
export * from './inputForm';
export * from './pagination';
