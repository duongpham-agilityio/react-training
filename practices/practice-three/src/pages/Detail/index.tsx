import {
  Box,
  Button,
  Center,
  Flex,
  HStack,
  Heading,
  IconButton,
  Image,
  Text,
  VStack,
} from '@chakra-ui/react';
import { memo } from 'react';

// Images
import ShoesImage from '@/assets/images/shoes.png';
import SocialImage from '@/assets/images/social.png';

// Icons
import { ArrowLeftOutline, ArrowRightOutline, Increase } from '@/assets/icons';

const Component = (): JSX.Element => {
  // Todo: Update handle to late

  return (
    <VStack
      h="full"
      backgroundImage={ShoesImage}
      backgroundPosition="center"
      backgroundSize="contain"
      backgroundRepeat="no-repeat"
      justifyContent="space-between"
      alignItems="unset"
    >
      {/* Info */}
      <Flex flex={1} minH={250} justifyContent="flex-end">
        <Box maxW={448}>
          <Heading
            fontSize={50}
            py={{
              base: 2,
              '2xl': 6,
            }}
            fontWeight="regular"
          >
            HUNK - Toffee
          </Heading>
          <Text
            fontSize={{
              base: 16,
              '2xl': 20,
            }}
            lineHeight={{
              base: 6,
              '2xl': 8,
            }}
            color="gray.20"
          >
            Add a spring to your step with this elegant and sleek pair of
            footwear! With its heavily padded ankle straps for secure footing
            and matted steel buckles that add flair, support and style go hand
            in hand with these hunk sandals
          </Text>
        </Box>
      </Flex>

      <Box h="fit-content">
        {/* Socials */}
        <Center justifyContent="space-between">
          <Text
            transform="rotate(-90deg)"
            textTransform="uppercase"
            fontWeight="regular"
            fontSize={16}
            letterSpacing={2}
          >
            Scroll down
          </Text>

          <Image src={SocialImage} />
        </Center>

        {/* Add to cart */}
        <HStack
          py={{
            base: 9,
            '2xl': 16,
          }}
          px={{
            base: 'unset',
            '2xl': 20,
          }}
          justifyContent="space-between"
        >
          {/* Slide control */}
          <Center>
            <Button
              height="fit-content"
              color="black"
              display="flex"
              flexDirection="column"
            >
              <Text fontSize={18} py={2} fontWeight="regular">
                Prev
              </Text>
              <ArrowLeftOutline />
            </Button>

            <Button
              height="fit-content"
              color="black"
              display="flex"
              flexDirection="column"
            >
              <Text fontSize={18} py={2} fontWeight="regular">
                Next
              </Text>
              <ArrowRightOutline />
            </Button>
          </Center>

          <Flex alignItems="center" gap={5}>
            <IconButton
              aria-label="Button play video"
              bg="gray.100"
              borderRadius="full"
              size="lg"
              icon={<Increase />}
            />
            <Text fontSize={18} fontWeight="regular" color="gray.30">
              Play video
            </Text>
          </Flex>

          <Button
            color="black"
            textTransform="uppercase"
            _hover={{
              color: 'gray.20',
            }}
            p="unset"
          >
            add to cart â€” $1599
          </Button>
        </HStack>
      </Box>
    </VStack>
  );
};

const Detail = memo(Component);

export default Detail;
