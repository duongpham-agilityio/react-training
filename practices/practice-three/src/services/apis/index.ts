export * from './account';
export * from './product';
export * from './image';
