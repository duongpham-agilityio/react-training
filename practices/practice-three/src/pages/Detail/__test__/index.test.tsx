import { fireEvent, render } from '@testing-library/react';
import {
  QueryClient,
  QueryClientProvider,
  useQuery,
} from '@tanstack/react-query';
import { BrowserRouter } from 'react-router-dom';

// Components
import Detail from '..';

// Mocks
import { products } from '@/mocks';

// Constants
import { MESSAGES } from '@/constants';
import { productAPI } from '@/services/apis';
import { IProduct } from '@/interface';

jest.mock('@/assets/images/shoes.png', () => ({
  default: '@/assets/images/shoes.png',
}));
jest.mock('@/assets/images/social.png', () => ({
  default: '@/assets/images/social.png',
}));

const queryClient: QueryClient = new QueryClient();

jest.mock('@tanstack/react-query', () => ({
  ...jest.requireActual('@tanstack/react-query'),
  useQuery: jest.fn(),
}));

const setup = () =>
  render(
    <BrowserRouter>
      <QueryClientProvider client={queryClient}>
        <Detail />
      </QueryClientProvider>
    </BrowserRouter>,
  );

describe('Details page', () => {
  it('With Loading', () => {
    (useQuery as jest.Mock).mockReturnValue({
      isLoading: true,
    });
    const { getByText } = setup();

    expect(getByText('Loading...')).toBeDefined();
  });

  it('With Error', () => {
    (useQuery as jest.Mock).mockReturnValue({
      isError: true,
    });
    const { getByText } = setup();

    expect(getByText(MESSAGES.FAIL_TO_FETCH)).toBeDefined();
  });
  it('Match to snapshot', () => {
    (useQuery as jest.Mock).mockReturnValue({
      isLoading: false,
      data: products[0],
    });
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it('Handle add product (valid)', () => {
    jest.spyOn(productAPI, 'getById').mockResolvedValue(products[0]);
    (useQuery as jest.Mock).mockReturnValue({
      isLoading: false,
      data: products[0],
    });

    const { container } = setup();
    const buttons = container.querySelectorAll('button');
    const button = buttons[buttons.length - 1];
    const mockEvent = jest.fn();

    button.addEventListener('click', mockEvent);

    fireEvent.click(button);

    expect(mockEvent).toBeCalled();
  });

  it('Handle add product (undefine)', () => {
    jest
      .spyOn(productAPI, 'getById')
      .mockResolvedValue(undefined as unknown as IProduct);
    (useQuery as jest.Mock).mockReturnValue({
      isLoading: false,
      data: products[0],
    });

    const { container } = setup();
    const buttons = container.querySelectorAll('button');
    const button = buttons[buttons.length - 1];
    const mockEvent = jest.fn();

    button.addEventListener('click', mockEvent);

    fireEvent.click(button);
  });

  it('Handle add product reject', () => {
    jest
      .spyOn(productAPI, 'getById')
      .mockRejectedValue(undefined as unknown as IProduct);
    (useQuery as jest.Mock).mockReturnValue({
      isLoading: false,
      data: products[0],
    });

    const { container } = setup();
    const buttons = container.querySelectorAll('button');
    const button = buttons[buttons.length - 1];
    const mockEvent = jest.fn();

    button.addEventListener('click', mockEvent);

    fireEvent.click(button);
  });
});
