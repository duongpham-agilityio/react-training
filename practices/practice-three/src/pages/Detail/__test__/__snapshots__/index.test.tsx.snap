// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details page Match to snapshot 1`] = `
<div>
  <div
    class="chakra-stack css-f25ded"
  >
    <div
      class="css-t3y40c"
    >
      <div
        class="css-1rr4qq7"
      />
      <div
        class="css-13n95wf"
      >
        <h2
          class="chakra-heading css-46oz1i"
        >
          HUNK - Toffee
        </h2>
        <p
          class="chakra-text css-1gsq3tq"
        >
          Add a spring to your step with this elegant and sleek pair of footwear! With its heavily padded ankle straps for secure footing and matted steel buckles that add flair, support and style go hand in hand with these hunk sandals
        </p>
      </div>
    </div>
    <div
      class="css-1sy0xge"
    >
      <div
        class="css-1lekzkb"
      >
        <p
          class="chakra-text css-gbfwt9"
        >
          Scroll down
        </p>
        <img
          class="chakra-image css-0"
          src="@/assets/images/social.png"
        />
      </div>
      <div
        class="chakra-stack css-32h9xz"
      >
        <div
          class="css-gmuwbf"
        >
          <button
            class="chakra-button css-1qqt8hf"
            type="button"
          >
            <p
              class="chakra-text css-s9vxsi"
            >
              Prev
            </p>
            <svg
              fill="none"
              height="12"
              width="59"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M59 6H5"
                stroke="#000"
                stroke-width="2"
              />
              <path
                d="M6 0v12L0 6l6-6Z"
                fill="#000"
              />
            </svg>
          </button>
          <button
            class="chakra-button css-1qqt8hf"
            type="button"
          >
            <p
              class="chakra-text css-s9vxsi"
            >
              Next
            </p>
            <svg
              fill="none"
              height="12"
              width="59"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 6h54"
                stroke="#000"
                stroke-width="2"
              />
              <path
                d="M53 12V0l6 6-6 6Z"
                fill="#000"
              />
            </svg>
          </button>
        </div>
        <div
          class="css-1dnaxzp"
        >
          <button
            aria-label="Button play video"
            class="chakra-button css-14144jg"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="#000"
              focusable="false"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m12.14 8.753-5.482 4.796c-.646.566-1.658.106-1.658-.753V3.204a1 1 0 0 1 1.659-.753l5.48 4.796a1 1 0 0 1 0 1.506z"
              />
            </svg>
          </button>
          <p
            class="chakra-text css-11kk9eu"
          >
            Play video
          </p>
        </div>
        <button
          class="chakra-button css-64vxoc"
          type="button"
        >
          add to cart â€” $1599
        </button>
      </div>
    </div>
  </div>
</div>
`;
